recipes/timers-and-such/direct/train_with_wav2vec.py

```diff
@@ -0,0 +1,432 @@
+        sig = sb.dataio.dataio.read_audio(wav)
+        return sig
+
+    sb.dataio.dataset.add_dynamic_item(datasets, audio_pipeline)
+
+    # 3. Define text pipeline:
+    @sb.utils.data_pipeline.takes("semantics")
+    @sb.utils.data_pipeline.provides(
+        "semantics", "token_list", "tokens_bos", "tokens_eos", "tokens"
+    )
+    def text_pipeline(semantics):
+        yield semantics
+        tokens_list = tokenizer.encode_as_ids(semantics)
+        yield tokens_list
+        tokens_bos = torch.LongTensor([hparams["bos_index"]] + (tokens_list))
+        yield tokens_bos
+        tokens_eos = torch.LongTensor(tokens_list + [hparams["eos_index"]])
+        yield tokens_eos
+        tokens = torch.LongTensor(tokens_list)
+        yield tokens
+
+    sb.dataio.dataset.add_dynamic_item(datasets, text_pipeline)
+
+    # 4. Set output:
+    sb.dataio.dataset.set_output_keys(
+        datasets,
+        ["id", "sig", "semantics", "tokens_bos", "tokens_eos", "tokens"],
+    )
+    return (
+        train_data,
+        valid_data,
+        test_real_data,
+        test_synth_data,
+        all_real_data,
+        tokenizer,
+    )
+
+
+if __name__ == "__main__":
+
+    # Load hyperparameters file with command-line overrides
+    hparams_file, run_opts, overrides = sb.parse_arguments(sys.argv[1:])
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    show_results_every = 100  # plots results every N iterations
+
+    # If distributed_launch=True then
+    # create ddp_group with the right communication protocol
+    sb.utils.distributed.ddp_init_group(run_opts)
+
+    # Create experiment directory
+    sb.create_experiment_directory(
+        experiment_directory=hparams["output_folder"],
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
+    # Dataset prep (parsing TAS)
+    from prepare import prepare_TAS  # noqa
+
+    # multi-gpu (ddp) save data preparation
+    run_on_main(
+        prepare_TAS,
+        kwargs={
+            "data_folder": hparams["data_folder"],
+            "save_folder": hparams["output_folder"],
+            "train_splits": hparams["train_splits"],
+            "type": "direct",
+            "skip_prep": hparams["skip_prep"],
+        },
+    )
+
+    # here we create the datasets objects as well as tokenization and encoding
+    (
+        train_set,
+        valid_set,
+        test_real_set,
+        test_synth_set,
+        all_real_set,
+        tokenizer,
+    ) = dataio_prepare(hparams)
+
+    # We download and pretrain the tokenizer
+    run_on_main(hparams["pretrainer"].collect_files)
+    hparams["pretrainer"].load_collected(device=run_opts["device"])
+
+    hparams["wav2vec2"] = hparams["wav2vec2"].to("cuda:0")
+
+    # freeze the feature extractor part when unfreezing
+    if not hparams["freeze_wav2vec"] and hparams["freeze_wav2vec_conv"]:
+        hparams["wav2vec2"].model.feature_extractor._freeze_parameters()
```

Comment:
I think we should move this into the wav2vec lobe..
@TParcollet , what do you think ?
It appear that freezing the conv and fine tuning the Transformer part of wav2vec gives the best results on the SLU..